plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java

```diff
@@ -87,36 +89,36 @@ public void importOrderFile(Object importOrderFile) {
 	}
 
 	public GrEclipseConfig greclipse() {
-		return greclipse(GrEclipseFormatterStep.defaultVersion());
+		return new GrEclipseConfig(null, this);
 	}
 
 	public GrEclipseConfig greclipse(String version) {
 		return new GrEclipseConfig(version, this);
 	}
 
 	public static class GrEclipseConfig {
-		final String version;
-		Object[] configFiles;
-		final FormatExtension extension;
+		private final EclipseConfiguration config;
+		private final FormatExtension extension;
 
-		GrEclipseConfig(String version, FormatExtension extension) {
+		GrEclipseConfig(@Nullable String version, FormatExtension extension) {
```

Comment:
null parameters that get replaced with defaults are more fragile than explicitly passing default values.  Consider:


```java
public static final String DEFAULT_VERSION = "1";

public static int computeFromVersion(String version) {
    return version.hashCode()
}

public static void main(String[] args) {
     computeFromVersion(DEFAULT_VERSION);
}
```

vs

```java
public static final String DEFAULT_VERSION = "1";

public static int computeFromVersion(@Nullable String version) {
    if (version == null) {
        version = DEFAULT_VERSION;
    }
    return version.hashCode()
}

public static void main(String[] args) {
     computeFromVersion(null);
}
```

The `null` case adds 3 lines, and what version ends up being is invisible to the user - they have to know the guts of the function to guess what might happen.  It also means that `computeFromVersion` is no longer a pure function of its inputs, it is now a function of its input and some hidden default state.

It's a small change, but multiplied across many lines and many programmers, null as default is costly.

Reply:
First of all, the constructor is package protected and not used by the user directly.

The EclipseConfiguration/EclipseBasedStepBuilder should be initialized by the step and prevent an invalid state.

What about
```java
GrEclipseConfig(FormatExtension extension) {
  this.extension = extension;
 config = GrEclipseFormatterStep.createConfig(GradleProvisioner.fromProject(extension.getProject()));
  extension.addStep(GrEclipseFormatterStep.createStep(config));
}

GrEclipseConfig(String version, FormatExtension extension)
{
  this(extension);
  config.setVersion(version);
}
```

The drawback here is that in the second constructor extension.replaceStep is required. But I find this concept anyway annoying. Gradle and Maven are both state machines ensuring that the execution is triggered after the configuration has been completed. Hence I would prefer to pass a factory method to FormatterStep. But I wanted to keep such things out of this PR.

Reply:
You've done a ton of work, and you're getting lots of style nitpicks which must be frustrating :)  Let me know when you're at a stopping point, so I can make some cleanup commits.  Might be easier to just show what I mean.  You'll be free to revert whichever of the commits you don't like.

Reply:
> You've done a ton of work, and you're getting lots of style nitpicks which must be frustrating :) 

@fvgh Indeed, so apologies for only adding to the potential frustration!

I'm impressed by the amount of work you've put into this PR, so please keep up the good work. I'll do my best to point out the _good_ things you do as well in the future, rather than just focusing on what could be improved. :)

Reply:
Sorry, I was short at time yesterday (and for the rest of the week). I think my formulations were a little bit harsh. That was not intended. I appreciate what I can learn from you guys. Feel free to take over.

Reply:
You weren't being harsh, I just know how I feel when my PR's get picked apart :)  I pushed up four commits.  Revert any that you don't think are helpful.  I see a few remaining problems:

- [ ] Overloading `setVersion` to mean either `setVersion` or `setHotfixCoordinatesUrl` is dangerous, better to split them up
- [ ] If we split it up, then we don't need to pass in the `supportedEclipseVersions`, we could just throw "version '123' is unsupported" if we don't find the resource
- [ ] The null default version is still dangerous, the steps should define a default version that we pass explicitly
- [ ] The resource handling code doesn't run on Windows, and 15 minutes of fiddling didn't fix it

I think fixing resource handling on Windows will be easier after the first three checkboxes are ticked.

As an aside, I'm less convinced than ever about `SemanticVersion` and `MavenCoordinates` vs `List<String>`, but I'll let it be since it's an implementation detail that we can fix later, and I might not be understanding the full picture :)